shared/src/main/scala-2.10/scalajson.ast/package.scala

```diff
@@ -343,4 +345,280 @@ package object ast {
       }
     } else pa == pb
   }
+
+  @inline private[ast] def radix: Int = 10
+
+  private[ast] def toInt(value: String): Option[Int] = {
+    @inline def maxLengthConstant: Int = 10
+    var limit: Int = -Integer.MAX_VALUE
+    var decimalFlag = false
+    var result: BigInt = 0
+    var negative = false
+    var multmin: Int = 0
+    var char: Char = 0
+    var i = 0
+    var eFlag = false
+    var trailingZeroes: BigInt = 0
+    var negativeEFlag: Boolean = false
+    var resultNegativeEFlag: BigInt = 0
+    var digitLength: BigInt = 0
+    multmin = limit / radix
+    if (value.charAt(0) == '-') {
+      limit = Integer.MIN_VALUE
+      negative = true
+      i += 1
+    }
+
+    while (i < value.length) {
+      char = value.charAt(i)
+      if (char == '.')
+        decimalFlag = true
+      else if ((char | 0x20) == 'e') {
+        eFlag = true
+        val charNext = value.charAt(i + 1)
+        if (charNext == '-')
+          negativeEFlag = true
+        if (negativeEFlag || charNext == '+') {
+          i += 1
+        }
+      } else {
+        if (!(eFlag || decimalFlag))
+          digitLength += 1
+
+        val digit = Character.digit(char, radix)
+
+        if (digit == 0)
+          if (!decimalFlag)
+            trailingZeroes += 1
+          else if (!decimalFlag)
+            trailingZeroes = 0
+
+        if (digit != 0 && (decimalFlag || eFlag))
+          if (!negativeEFlag)
+            return None
+          else {
+            if (trailingZeroes != 0) {
+              resultNegativeEFlag *= radix
+              resultNegativeEFlag += digit
+              if (trailingZeroes >= resultNegativeEFlag) {
+                var i2: Int = 0
+                while (i2 < resultNegativeEFlag) {
+                  result = result / radix
+                  i2 += 1
+                }
+              } else
+                return None
+            }
+          }
+
+        val maxLenCheck = digitLength <= maxLengthConstant
+
+        if (result < multmin && maxLenCheck && !eFlag)
+          return None
+
+        if (!(digit == 0 && (decimalFlag || eFlag))) {
+
+          if (!negativeEFlag) {
+            result *= radix
+            if (result < limit + digit && maxLenCheck)
```

Comment:
Just count zeros with an `Int`.  You aren't going to have more than two billion because they won't fit in the string anyway, and likewise if the exponent isn't in that range you're not going to have a representable integer.
Then you need to detect when you're in danger of overflow and divide it into three cases based on leading digit: overflow impossible; overflow certain; and overflow possible.  When overflow is possible and actually happens, the sign of `result` will flip.  So you look for that.